caas/kubernetes/provider/metadata.go

```diff
@@ -39,5 +46,98 @@ func getCloudProviderFromNodeMeta(node core.Node) string {
 			return k
 		}
 	}
+	// TODO - add microk8s node label check when available
+	hostname, err := os.Hostname()
+	if err != nil {
+		return ""
+	}
+	hostLabel, _ := node.Labels["kubernetes.io/hostname"]
+	if node.Name == hostname && hostLabel == hostname {
+		return "microk8s"
+	}
 	return ""
 }
+
+// GetClusterMetadata implements ClusterMetadataChecker.
+func (k *kubernetesClient) GetClusterMetadata(storageClass string) (*caas.ClusterMetadata, error) {
+	var result caas.ClusterMetadata
+	var err error
+	result.Regions, err = k.listHostCloudRegions()
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot determine cluster region")
+	}
+
+	if storageClass != "" {
+		sc, err := k.StorageV1().StorageClasses().Get(storageClass, v1.GetOptions{IncludeUninitialized: true})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return nil, errors.Trace(err)
+		}
+		if err == nil {
+			result.NominatedStorageClass = &caas.StorageProvisioner{
+				Provisioner: sc.Provisioner,
+				Parameters:  sc.Parameters,
+			}
+		}
+	} else {
+		storageClasses, err := k.StorageV1().StorageClasses().List(v1.ListOptions{})
+		if err != nil {
+			return nil, errors.Annotate(err, "listing storage classes")
+		}
+		for _, sc := range storageClasses.Items {
+			if v, ok := sc.Annotations["storageclass.kubernetes.io/is-default-class"]; ok && v != "false" {
+				result.NominatedStorageClass = &caas.StorageProvisioner{
+					Provisioner: sc.Provisioner,
+					Parameters:  sc.Parameters,
+				}
+				break
+			}
+		}
+	}
+	return &result, nil
+}
+
+const regionLabelName = "failure-domain.beta.kubernetes.io/region"
+
+// listHostCloudRegions lists all the cloud regions that this cluster has worker nodes/instances running in.
+func (k *kubernetesClient) listHostCloudRegions() (set.Strings, error) {
+	// we only check 5 worker nodes as of now just run in the one region and
+	// we are just looking for a running worker to sniff its region.
+	nodes, err := k.CoreV1().Nodes().List(v1.ListOptions{Limit: 5})
+	if err != nil {
+		return nil, errors.Annotate(err, "listing nodes")
+	}
+	result := set.NewStrings()
+	for _, n := range nodes.Items {
+		var cloudRegion, v string
+		var ok bool
+		if v = getCloudProviderFromNodeMeta(n); v == "" {
+			continue
+		}
+		cloudRegion += v
+		if v, ok = n.Labels[regionLabelName]; ok && v != "" {
+			cloudRegion += "/" + v
+		}
```

Comment:
So in `<cloud>/<region>`, both `<cloud>` and `<region>` were required. Do we allow here can only return `<cloud>` without `<region>` now?

Reply:
Yes. Some clouds don't have regions yet, like microk8s. We only need the region for JAAS. But we always need the cloud so we can see what the storage needs to be.

Reply:
ic thx